### PR TITLE
Stringified annotations

### PR DESCRIPTION
**Pull request by azidar**
Restricts annotations to be string-based (and thus less typesafe)
Makes annotations more easily serializable and interact with Chisel

### CODE REVIEW

File: src/test/scala/firrtlTests/DriverSpec.scala [hunk 1]
```diff
@@ -91,8 +93,9 @@ class DriverSpec extends FreeSpec with Matchers {
 
         val firrtlOptions = optionsManager.firrtlOptions
         firrtlOptions.annotations.length should be (3)
+        val x = classOf[InlineInstances]
```
Comment (chick) [inline src/test/scala/firrtlTests/DriverSpec.scala]: Looks like x here is never used.

File: src/test/scala/firrtlTests/DriverSpec.scala [hunk 2]
```diff
@@ -4,10 +4,12 @@ package firrtlTests
 
 import java.io.File
 
-import firrtl.passes.memlib.{InferReadWriteAnnotation, ReplSeqMemAnnotation}
+import firrtl.Annotations.Annotation
```
Comment (chick) [inline src/test/scala/firrtlTests/DriverSpec.scala]: This import does not appear to be used

**Comment by chick** [comment]
reading in annotations has an implementation in the firrtl add-dedup branch.  Is add-dedup going anywhere or should I break out into separate PR.  (Contrary to what I said earlier @azidar, I did have some annotation work pending in firrtl)

**Comment by azidar** [comment]
I thought add-dedup was already merged (or maybe just the commit that actually added the dedup transform). Regardless, @chick I'll go ahead and take a look at your code in add-dedup, and add the related commits here.

**Comment by jackkoenig** [comment]
I didn't include the reading in annotations stuff when I merged parts of
add-dedup because they didn't make sense without the changes in this PR.

**Comment by azidar** [comment]
@jackkoenig is this ok to merge?

**Comment by jackkoenig** [comment]
I've been in meetings all day so I haven't had a chance to really review it

**Comment by azidar** [comment]
Oh no worries, I was just confused about your earlier request for changes, so just didn't know what the status was.

File: src/main/scala/firrtl/passes/clocklist/ClockListTransform.scala [hunk 3]
```diff
@@ -65,7 +56,7 @@ class ClockListTransform extends Transform {
   def passSeq(top: String, writer: Writer): Seq[Pass] =
     Seq(new ClockList(top, writer))
   def execute(state: CircuitState): CircuitState = getMyAnnotations(state) match {
-    case Seq(ClockListAnnotation(ModuleName(top, CircuitName(state.circuit.main)), out)) => 
+    case Seq(Annotation(ModuleName(top, CircuitName(state.circuit.main)), _, out)) =>
```
Comment (jackkoenig) [inline src/main/scala/firrtl/passes/clocklist/ClockListTransform.scala]: For the record, it isn't necessary but you could add an unapply to ClockListAnnotation to be able to keep the same unapply.

File: src/main/scala/firrtl/passes/memlib/InferReadWrite.scala [hunk 4]
```diff
@@ -155,6 +153,6 @@ class InferReadWrite extends Transform with PassBased {
   )
   def execute(state: CircuitState): CircuitState = getMyAnnotations(state) match {
     case Nil => CircuitState(state.circuit, state.form)
-    case Seq(InferReadWriteAnnotation(_)) => CircuitState(runPasses(state.circuit), state.form)
+    case Seq(Annotation(_, _, _)) => CircuitState(runPasses(state.circuit), state.form)
```
Comment (jackkoenig) [inline src/main/scala/firrtl/passes/memlib/InferReadWrite.scala]: Same unapply idea as above

File: src/main/scala/firrtl/passes/memlib/ReplaceMemTransform.scala [hunk 5]
```diff
@@ -82,18 +82,18 @@ Optional Arguments:
...
+    val outputConfig = passOptions.getOrElse(
+      OutputConfigFileName, 
+      error("No output config file provided for ReplSeqMem!" + usage)
+    )
+    val passCircuit = passOptions.getOrElse(
+      PassCircuitName, 
+      error("No circuit name specified for ReplSeqMem!" + usage)
+    )
+    val target = CircuitName(passCircuit)
+    Annotation(target, classOf[ReplSeqMem], t)
```
Comment (jackkoenig) [inline src/main/scala/firrtl/passes/memlib/ReplaceMemTransform.scala]: Would probably be better with an apply method on ReplSeqMem that just accepts target and t (similar to previous comments)

File: src/main/scala/firrtl/passes/memlib/ReplaceMemTransform.scala [hunk 6]
```diff
@@ -139,7 +139,7 @@ class ReplSeqMem extends Transform with SimpleRun {
     getMyAnnotations(state) match {
       case Nil => state.copy(annotations = None) // Do nothing if there are no annotations
       case p => (p.collectFirst { case a if (a.target == CircuitName(state.circuit.main)) => a }) match {
-        case Some(ReplSeqMemAnnotation(t)) =>
+        case Some(Annotation(_, _, t)) =>
```
Comment (jackkoenig) [inline src/main/scala/firrtl/passes/memlib/ReplaceMemTransform.scala]: Same comment as before on unapply

File: src/main/scala/firrtl/passes/wiring/WiringTransform.scala [hunk 7]
```diff
@@ -64,20 +55,17 @@ class WiringTransform extends Transform with SimpleRun {
...
-            comp(pin) = c.name
-          case TopAnnotation(m, pin) => tops(pin) = m.name
-        }
+      p.foreach { 
+        case Annotation(m, _, SinkAnnotation.matcher(pin)) => sinks(pin) = sinks.getOrElse(pin, Set.empty) + m.name
+        case Annotation(ComponentName(n, ModuleName(m, CircuitName(c))), _, SourceAnnotation.matcher(pin)) =>
+          sources(pin) = m
+          comp(pin) = n
+        case Annotation(m, _, TopAnnotation.matcher(pin)) => tops(pin) = m.name
       }
```
Comment (jackkoenig) [inline src/main/scala/firrtl/passes/wiring/WiringTransform.scala]: last comment like this but yeah unapplies would make this better

File: src/main/scala/firrtl/Driver.scala [hunk 8]
```diff
@@ -2,15 +2,19 @@
...
 import logger.Logger
 
 import scala.io.Source
-import Annotations._
+import annotations._
+import firrtl.annotations.AnnotationYamlProtocol._
 
 import Parser.{InfoMode, IgnoreInfo}
 import scala.collection._
+import net.jcazevedo.moultingyaml._
```
Comment (azidar) [inline src/main/scala/firrtl/Driver.scala]: done

File: src/main/scala/firrtl/Driver.scala [hunk 9]
```diff
@@ -72,6 +76,28 @@ object Driver {
...
+    /*
+     Right now annotations will be looked for always based on the
+     s"$targetDirName/$topName.anno" or s"$annotationFileNameOverride.anno"
+     If found they will be added to the annotations already in the
+     optionsManager.firrtlOptions, duplicates may be created, but this should be ok
+    */
+    val firrtlConfig = optionsManager.firrtlOptions
+    val annotationFileName = firrtlConfig.getAnnotationFileName(optionsManager)
+    val annotationFile = new File(annotationFileName)
+    if(annotationFile.exists()) {
```
Comment (azidar) [inline src/main/scala/firrtl/Driver.scala]: done

File: src/main/scala/firrtl/ExecutionOptionsManager.scala [hunk 10]
```diff
@@ -215,8 +225,17 @@ trait HasFirrtlOptions {
...
+  }
+
+  parser.opt[String]("annotation-file")
+    .abbr("a")
+    .valueName ("<output>").
+    foreach { x =>
+      firrtlOptions = firrtlOptions.copy(outputFileNameOverride = x)
+    }.text {
+    "use this to override the default annotation file name, default is empty"
+  }
```
Comment (azidar) [inline src/main/scala/firrtl/ExecutionOptionsManager.scala]: Its something to think about, but I think this should be in FirrtlOptions, at least until we think chisel should accept an annotation file as a argument (which we haven't planned on yet).

File: src/main/scala/firrtl/annotations/Annotations.scala [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+  val transformClass: String = transform.getName
+  def serialize: String = this.toString
+  def update(tos: Seq[Named]): Seq[Annotation] = {
+    check(target, tos, this)
+    propagate(target, tos, duplicate)
+  }
+  def propagate(from: Named, tos: Seq[Named], dup: Named=>Annotation): Seq[Annotation] = tos.map(dup(_))
+  def check(from: Named, tos: Seq[Named], which: Annotation): Unit = {}
+  def duplicate(n: Named) = new Annotation(n, transform, value)
+}
```
Comment (azidar) [inline src/main/scala/firrtl/annotations/Annotations.scala]: done

File: src/main/scala/firrtl/annotations/AnnotationUtils.scala [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+          buildup(SubField(e, tokens(1), UnknownType), tokens.drop(2))
+        case Nil => e
+      }
+      val root = Reference(tokens.head, UnknownType)
+      buildup(root, tokens.tail)
+    }
+    if(validComponentName(s)) {
+      parse(tokenize(s))
+    } else error(s"Cannot convert $s into an expression.")
+  }
```
Comment (azidar) [inline src/main/scala/firrtl/annotations/AnnotationUtils.scala]: done

File: src/main/scala/firrtl/passes/clocklist/ClockListTransform.scala [hunk 13]
```diff
@@ -16,17 +16,8 @@ import memlib.AnalysisUtils._
...
-  def transform = classOf[ClockListTransform]
-  def duplicate(n: Named) = n match {
-    case m: ModuleName => this.copy(target = m, outputConfig = outputConfig)
-    case _ => error("Clocklist can only annotate a module.")
-  }
-}
-
 object ClockListAnnotation {
-  def apply(t: String) = {
+  def apply(t: String): Annotation = {
```
Comment (azidar) [inline src/main/scala/firrtl/passes/clocklist/ClockListTransform.scala]: done

File: src/main/scala/firrtl/passes/clocklist/ClockListTransform.scala (hunk 3 above)
Comment (azidar) [inline src/main/scala/firrtl/passes/clocklist/ClockListTransform.scala]: done

File: src/main/scala/firrtl/passes/memlib/InferReadWrite.scala (hunk 4 above)
Comment (azidar) [inline src/main/scala/firrtl/passes/memlib/InferReadWrite.scala]: done

File: src/main/scala/firrtl/passes/memlib/ReplaceMemTransform.scala (hunk 5 above)
Comment (azidar) [inline src/main/scala/firrtl/passes/memlib/ReplaceMemTransform.scala]: done

File: src/main/scala/firrtl/passes/memlib/ReplaceMemTransform.scala (hunk 6 above)
Comment (azidar) [inline src/main/scala/firrtl/passes/memlib/ReplaceMemTransform.scala]: done

File: src/main/scala/firrtl/passes/wiring/WiringTransform.scala (hunk 7 above)
Comment (azidar) [inline src/main/scala/firrtl/passes/wiring/WiringTransform.scala]: done

**Pull request merged**